`ultralytics 8.3.88` Solutions refactor and improvements

RizwanMunawar opened the pull request:
<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
--->


## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

**WARNING ⚠️** this PR is very large, summary may not cover all changes.

### 🌟 Summary
This PR introduces significant updates and enhancements to the Ultralytics documentation and solutions, focusing on improving usability, expanding functionality, and refining the user experience for various computer vision tasks.

### 📊 Key Changes
- **Documentation Enhancements**:
  - Added detailed guides for new solutions like `ObjectBlurrer`, `ObjectCropper`, `InstanceSegmentation`, and `VisionEye`.
  - Improved consistency and clarity in existing guides (e.g., `Heatmaps`, `Analytics`, `Queue Management`).
  - Introduced reusable macros for argument tables (`solutions-args`, `track-args`, `visualization-args`) to streamline documentation updates.
  - Added reference pages for new solution classes (`ObjectBlurrer`, `ObjectCropper`, `InstanceSegmentation`, `VisionEye`).

- **New Solutions**:
  - **ObjectBlurrer**: Enables real-time blurring of detected objects for privacy-focused applications.
  - **ObjectCropper**: Allows precise cropping of detected objects from images or videos.
  - **InstanceSegmentation**: Provides instance segmentation with mask visualization.
  - **VisionEye**: Simulates human-like object mapping and tracking with a focus point.

- **Solution Updates**:
  - Unified the `process` method across solutions for consistent API usage.
  - Enhanced `Analytics` and `Heatmap` solutions with better visualization and data handling.
  - Improved `AIGym` for workout monitoring with detailed keypoint tracking and angle calculations.

- **Testing and CLI Improvements**:
  - Updated tests to include new solutions and ensure compatibility.
  - Enhanced CLI support for solutions with additional arguments like `blur_ratio`, `crop_dir`, and `vision_point`.

### 🎯 Purpose & Impact
- **Improved Usability**: Simplified the integration and usage of solutions with consistent APIs and detailed documentation.
- **Expanded Functionality**: Introduced new solutions to address diverse use cases like privacy, segmentation, and object mapping.
- **Enhanced User Experience**: Streamlined CLI commands and added visual feedback for better interaction.
- **Developer Efficiency**: Reusable macros and unified methods reduce maintenance overhead and improve scalability.

UltralyticsAssistant commented:
👋 Hello @RizwanMunawar, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

This is an automated response, but don’t worry—one of our Ultralytics engineers will assist you soon! 😊

Laughing-q commented:
@RizwanMunawar seems like this PR is way more than `Annorator` to `SolutionAnnotator` update 😂

RizwanMunawar commented:
> @RizwanMunawar seems like this PR is way more than `Annorator` to `SolutionAnnotator` update 😂

@Laughing-q Yes, we can now call it `update solutions structure`. Initially, I was considering making minor changes, but I realized it would result in multiple PRs and could eventually impact users over time. So, it’s better to merge everything at once. :)

RizwanMunawar committed:
Merge bc6a5bb8cb78a2caa2a043aca8c245540e2fdf6d into 9d282ddeee5a5e4435fca7170e91e6ea27bf1911

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
Merge 3a0b4a28791ef3483b605619dced312f731e2c01 into aecf1da32b50e144b226f54ae4e979dc3cc62145

UltralyticsAssistant committed:
Auto-update Ultralytics Docs by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
update docs

RizwanMunawar committed:
update docs

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
Final update

RizwanMunawar committed:
Final update

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
Final update

RizwanMunawar committed:
update solutions tests

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar commented:
@Laughing-q Hi, finally this PR is almost ready :) Can you please review it this week sometime. Thanks

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
Optimize seg_bbox calculations

RizwanMunawar committed:
resolve conflicts with main

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
remove solutions plotting from plotting.py

RizwanMunawar committed:
remove solutions plotting from plotting.py

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

Laughing-q committed:
Merge branch 'main' into annotator-overrides-for-solutions

Laughing-q commented:
@RizwanMunawar well take a look this night

Laughing-q commented on docs/en/guides/instance-segmentation-and-tracking.md:
````diff
@@ -41,96 +41,59 @@ There are two types of instance segmentation tracking available in the Ultralyti
         ```python
````
@RizwanMunawar we can remove this `Instance Segmentation` title since there's only one demo now.

Laughing-q commented on ultralytics/solutions/object_blurrer.py:
```diff
@@ -0,0 +1,78 @@
+
+        self.blur_ratio = int(self.CFG["blur_ratio"] * 100)  # Blur intensity multiplier
+
+    def blur(self, im0):
```
probably add `blur_ratio` as an argument of `blur` method instead of introducing as a global argument

Laughing-q replied:
same as the new `vision_point`

Laughing-q commented on ultralytics/solutions/security_alarm.py:
```diff
@@ -141,4 +141,7 @@ def monitor(self, im0):
 
-        return im0  # return output image for more usage
+        # return output dictionary with summary for more usage
+        return SolutionResults(im0=im0, total_tracks=len(self.track_ids), email_sent=self.email_sent).summary(
```
is this calling `summary`(which returns a dict) everywhere necessary? isn't it more convenient that we directly use the `SolutionResults` object hence access attributes by dot? 
`result["im0"]`  --> `result.im0`

Laughing-q replied:
also I figured the `im0` here is the gets plotted, then probably rename it to something `plot_im`, since in ultralytics package we use `im0` represents original basically everywhere.

RizwanMunawar replied:
@Laughing-q im0 is replaced with `plot_im` in [2a89f8c](https://github.com/ultralytics/ultralytics/pull/18491/commits/2a89f8c5b3f1158231a31e66ec08945247d1b445).

RizwanMunawar replied:
@Laughing-q removed the summary function from `SolutionResults`, better in terms of usage now. [c1fe2df](https://github.com/ultralytics/ultralytics/pull/18491/commits/c1fe2df2287542757e3c18bf1151a362517bc40a)

Laughing-q commented:
@RizwanMunawar Let's move all the headlines that only have one example:
<img width="620" alt="屏幕截图 2025-02-06 234344" src="https://github.com/user-attachments/assets/c852b2a6-e66f-4dbf-a32c-3c37b86752b2" />
And we need more information in our docs about the `SolutionResults` that gathers all the returned solution results.

RizwanMunawar committed:
remove heading for solutions having single example

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
Remove vision_point and blur_ratio from default.yaml

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

RizwanMunawar committed:
Merge bc27f3900fb9a45a0fa648ab13b209a4c97f87af into c1860b833361c9f88e53e522c3b8530e5ff698a6

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar commented:
@Laughing-q Yes that's a good idea. 

- I have removed all the headings from the Solutions documentation that have a single example in [8444fda](https://github.com/ultralytics/ultralytics/pull/18491/commits/8444fda8dd75799d458246998204d081410e382a).
- `blur_ratio` and `vision_point` are currently global, but your idea is also ok. I have updated this in [ed23960](https://github.com/ultralytics/ultralytics/pull/18491/commits/ed2396089ee8735efd2f43368621d9e46c4c2892).

- I will replace `im0` with `plot_im` and also explore `SolutionsResults` If we can use some other way around tomorrow. Then I think this PR will be ready :)

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
Add SolutionResults details in index.md and update im0 with plot_im

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
Merge b1b518dbfb6fba26f946259f784b1f3bb17682ca into 580b4766c52c1404b6ded6f88860b2ad13117236

UltralyticsAssistant committed:
Auto-update Ultralytics Docs by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
update

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
Merge 7a777deec43076400e25ae96ac2b3b7efaa99397 into 580b4766c52c1404b6ded6f88860b2ad13117236

UltralyticsAssistant committed:
Auto-update Ultralytics Docs by https://ultralytics.com/actions

RizwanMunawar committed:
replace results['im0'] with results.imo

RizwanMunawar committed:
replace results['im0'] with results.imo

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
remove summary from SolutionResults

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update index.md

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

Laughing-q committed:
Merge branch 'main' into annotator-overrides-for-solutions

Laughing-q commented on ultralytics/solutions/object_cropper.py:
```diff
@@ -0,0 +1,88 @@
+        """
+        super().__init__(**kwargs)
+
+        self.crop_directory = "cropped-detections"  # Directory for storing cropped detections
```
@RizwanMunawar I tested this argument by 
```python
cropper = solutions.ObjectCropper(
    show=True,  # display the output
    model="yolo11n.pt",  # model for object cropping i.e yolo11x.pt.
    classes=[0, 2],  # crop specific classes i.e. person and car with COCO pretrained model.
    # conf=0.5  # adjust confidence threshold for the objects.
    crop_directory="crop_directory"
)
```
but it does not work, it still uses this default `cropped-detections`.

Laughing-q replied:
@RizwanMunawar any updates on this one?

RizwanMunawar replied:
@Laughing-q the crop_dir arg should work fine now. [172a930](https://github.com/ultralytics/ultralytics/pull/18491/commits/172a930033d36f6aacfb2f0056ea1674b7698bcb)

Laughing-q commented on ultralytics/solutions/object_counter.py:
```diff
@@ -172,8 +174,9 @@ def count(self, im0):
-        self.annotator = Annotator(im0, line_width=self.line_width)  # Initialize annotator
         self.extract_tracks(im0)  # Extract tracks
+        plot_im = im0  # For plotting the results
+        self.annotator = SolutionAnnotator(plot_im, line_width=self.line_width)  # Initialize annotator
```
@RizwanMunawar why we make it `self.annotator`? for now I only see annotator used in this `count` function. BTW I saw some other solutions this `annotator` is not a self attribute.

Laughing-q commented on ultralytics/solutions/object_cropper.py:
```diff
@@ -0,0 +1,88 @@
+        for box, cls in zip(boxes, clss):
+            self.crop_idx += 1
+            crop_obj = plot_im[int(box[1]) : int(box[3]), int(box[0]) : int(box[2])]  # Crop the detected object
+            cv2.imwrite(os.path.join(self.crop_directory, f"crop-{self.crop_idx}.jpg"), crop_obj)  # Save cropped image
```
@RizwanMunawar can we reuse our current `save_one_box` function here?

Laughing-q commented on ultralytics/solutions/solutions.py:
```diff
@@ -173,6 +184,536 @@ def display_output(self, im0):
         if self.CFG.get("show") and self.env_check:
-            cv2.imshow("Ultralytics Solutions", im0)
+            cv2.imshow("Ultralytics Solutions", plot_im)
             if cv2.waitKey(1) & 0xFF == ord("q"):
```
@RizwanMunawar My initial thought for this is that I could use `q` to quit the cv2 display window, but it's not working as I expected, what's the `q` here for?

RizwanMunawar replied:
@Laughing-q Yes, 'q` is mainly used to quit the current frame, and it will now work as expected. resolved in [6392794](https://github.com/ultralytics/ultralytics/pull/18491/commits/639279467fbe8f9f7d65d791010021e06404c6ee)

Laughing-q replied:
@RizwanMunawar it does not work to me😂
```python
from ultralytics import solutions
cropper = solutions.ObjectCropper(
    show=True,  # display the output
    model="yolo11n.pt",  # model for object cropping i.e yolo11x.pt.
    classes=[0, 2],  # crop specific classes i.e. person and car with COCO pretrained model.
    # conf=0.5  # adjust confidence threshold for the objects.
    crop_directory="crop_directory"
)

# Process video
while cap.isOpened():
    success, im0 = cap.read()

    if not success:
        print("Video frame is empty or processing is complete.")
        break
    results = cropper.crop(im0)
cap.release()
cv2.destroyAllWindows()  # destroy all opened windows
```
using `q` does not quit and exit.

RizwanMunawar replied:
@Laughing-q really, on windows it's working fine :)

RizwanMunawar replied:
@Laughing-q the `crop_dir` arg should work fine now. [172a930](https://github.com/ultralytics/ultralytics/pull/18491/commits/172a930033d36f6aacfb2f0056ea1674b7698bcb)

Laughing-q commented on ultralytics/solutions/ai_gym.py:
```diff
@@ -82,14 +85,14 @@ def monitor(self, im0):
 
             # Initialize annotator
-            self.annotator = Annotator(im0, line_width=self.line_width)
+            self.annotator = SolutionAnnotator(plot_im, line_width=self.line_width)
```
@RizwanMunawar same annotator issue

RizwanMunawar replied:
@Laughing-q fixed in [9436ca5](https://github.com/ultralytics/ultralytics/pull/18491/commits/9436ca5058cb59f985c1093dd3d4a8c3e714ae16)

Laughing-q commented on ultralytics/solutions/object_cropper.py:
```diff
@@ -0,0 +1,88 @@
+            >>> summary = cropper.crop(frame)
+            >>> print(summary)
+        """
+        results = self.model.predict(im0, classes=self.classes, conf=self.conf, iou=self.iou, device="cpu")[0]
```
@RizwanMunawar no way to specify a device here? I saw other solutions we have `device` available for users.

RizwanMunawar replied:
@Laughing-q yes nice catch! resolved in [24c57ba](https://github.com/ultralytics/ultralytics/pull/18491/commits/24c57ba512bca377eecc8d87ec577f0473199ebe)

Laughing-q commented on ultralytics/solutions/speed_estimation.py:
```diff
@@ -58,15 +58,16 @@ def estimate_speed(self, im0):
-        self.annotator = Annotator(im0, line_width=self.line_width)  # Initialize annotator
         self.extract_tracks(im0)  # Extract tracks
+        plot_im = im0  # For plotting the results
+        self.annotator = SolutionAnnotator(plot_im, line_width=self.line_width)  # Initialize annotator
```
same here

RizwanMunawar replied:
@Laughing-q resolved in [2ddb768](https://github.com/ultralytics/ultralytics/pull/18491/commits/2ddb7681f9c8749ae6e393e3bf8a4dcbbcebdc2d)

Laughing-q commented on ultralytics/solutions/trackzone.py:
```diff
@@ -45,24 +45,28 @@ def trackzone(self, im0):
         """
-        self.annotator = Annotator(im0, line_width=self.line_width)  # Initialize annotator
+        plot_im = im0  # For plotting the results
+        self.annotator = SolutionAnnotator(plot_im, line_width=self.line_width)  # Initialize annotator
```
same annotator

RizwanMunawar replied:
@Laughing-q resolved in [e158355](https://github.com/ultralytics/ultralytics/pull/18491/commits/e15835573c097f420416c6b3670af61a8235375b)

Laughing-q commented on ultralytics/solutions/vision_eye.py:
```diff
@@ -0,0 +1,78 @@
+        super().__init__(**kwargs)
+        self.vision_point = (20, 20)
+
+    def set_vision_point(self, vision_point):
```
@RizwanMunawar I meant adding the `vision_point` as an argument for `mapping` method😂
```python
def mapping(im0, vision_point=(20. 20))
```

Laughing-q commented on ultralytics/solutions/object_blurrer.py:
```diff
@@ -0,0 +1,93 @@
+
+        self.blur_ratio = 50  # Blur intensity multiplier
+
+    def set_blur_ratio(self, blur_ratio=0.5):
```
@RizwanMunawar I meant adding the `blur_ratio` as an argument for `blur` method😂
```python
def blur(im0, blur_ratio=0.5)
```

RizwanMunawar replied:
@Laughing-q as discussed, let's use this function :)

Laughing-q commented:
@RizwanMunawar Added some reviews here, also I suggest we could have an `__call__` method in BaseSolution and apply it to every solution, for example for objectcounter we could directly call `counter(im0)`:
```python
r = solutions.ObjectCounter(
    show=True,  # display the output
    region=region_points,  # pass region points
    model="yolo11n.pt",  # model="yolo11n-obb.pt" for object counting with OBB model.
    verbose=False
    # classes=[0, 2],           # count specific classes i.e. person and car with COCO pretrained model.
    # tracker="botsort.yaml"    # Choose trackers i.e "bytetrack.yaml"
)
results = counter(im0)
```
```python
results = counter.count(im0)  # current way
```

RizwanMunawar committed:
fix device argument in object_cropper.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
remove solutions examples from index.md

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar commented on docs/en/solutions/index.md:
```diff
@@ -42,33 +42,771 @@ Here's our curated list of Ultralytics solutions that can be used to create awes
+
+`SolutionResults` objects have the following attributes:
+
+| Attribute            | Type              | Description                                       |
```
@Laughing-q resolved in [8b997a1](https://github.com/ultralytics/ultralytics/pull/18491/commits/8b997a111b27e8875878d0b9c39000afae66f19f) :)

RizwanMunawar committed:
remove set_vision_point function

RizwanMunawar committed:
remove self from annotator in speed_estimation.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
remove self from annotator in trackzone.py

RizwanMunawar committed:
remove self from annotator in ai_gym.py

RizwanMunawar committed:
update SolutionResults

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar commented on ultralytics/solutions/solutions.py:
```diff
@@ -173,6 +184,536 @@ def display_output(self, im0):
+
+        # Override with user-defined values
+        self.__dict__.update(kwargs)
+        if self.verbose:
```
@Laughing-q fixed in [ddb7d8d](https://github.com/ultralytics/ultralytics/pull/18491/commits/ddb7d8d71c65f510c88a74bff8728c39ee42f2a6)

RizwanMunawar committed:
update solutions.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Laughing-q commented on docs/en/guides/analytics.md:
```diff
@@ -86,30 +86,37 @@ This guide provides a comprehensive overview of three fundamental types of [data
-| `line_width`     | `int`  | `2`     | Line thickness for bounding boxes.                   |
-| `show`           | `bool` | `False` | Flag to control whether to display the video stream. |
-
-### Arguments `model.track`
```
@RizwanMunawar any updates on this one?

RizwanMunawar replied:
@Laughing-q resolved in [def8041](https://github.com/ultralytics/ultralytics/pull/18491/commits/def8041fdc9eaf1187bec7896f4659ef791c28f0).

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
Update object_cropper.py

RizwanMunawar committed:
update object-cropping.md

RizwanMunawar committed:
update vision-eye

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update object_blurrer.py

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update functions information in docs

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update solutions docs

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
update

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update analytics.md

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
Add __call__ to BaseSolutions

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
fix docs build

RizwanMunawar committed:
update test_solutions.py

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

RizwanMunawar committed:
update region_counter.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
remove self.annotator

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
update object_cropper.py

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
final-update

RizwanMunawar committed:
update object_cropper.py

RizwanMunawar committed:
update object_cropper.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Laughing-q committed:
Update object_cropper.py

Laughing-q committed:
Update index.md

Laughing-q committed:
Update object-counting.md

Laughing-q committed:
Update object-cropping.md

Laughing-q committed:
Update object-blurring.md

Laughing-q committed:
Update analytics.py

Laughing-q committed:
Update analytics.py

Laughing-q committed:
Update object_counter.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update ai_gym.py

RizwanMunawar committed:
update ai_gym.py

RizwanMunawar committed:
update ai_gym.py

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
update distance_calculation.py

RizwanMunawar committed:
update heatmap.py

RizwanMunawar committed:
update instance_segmentation.py

RizwanMunawar committed:
update object-blurring.md & object_blurrer.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update parking_management.py and parking-management.md

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update queue_management.py

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

RizwanMunawar committed:
update region_counter.py

RizwanMunawar committed:
update security_alarm.py

RizwanMunawar committed:
update speed_estimation.py

RizwanMunawar committed:
update trackzone.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update vision_eye.py

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

RizwanMunawar committed:
update object-cropping.md

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update region-counting.md

RizwanMunawar committed:
update security-alarm-system.md

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update instance-segmentation-and-tracking.md

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update vision-eye.md

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update distance-calculation.md

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update parking-management.md

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

RizwanMunawar committed:
update docs

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update region-counting.md

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update instance-segmentation-and-tracking.md

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

Laughing-q committed:
Merge branch 'main' into annotator-overrides-for-solutions

Laughing-q committed:
fix SolutionResults docstring

Laughing-q committed:
Update solutions-args.md

Laughing-q committed:
Update visualization-args.md

Laughing-q committed:
Update track-args.md

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Laughing-q committed:
Merge branch 'main' into annotator-overrides-for-solutions

Laughing-q commented:
@glenn-jocher This PR is ready to me. Here's a briefly summary of what @RizwanMunawar has done in this PR.
- There are a lot plottig functions in `Annotator` class but designed specifically for solutions so I suggested to have a specific `SolutionAnnotator` inherited from our ultralytics `Annotator`, and use this new `SolutionAnnotator` in solutions plotting.
- All the solutions used to return the plotted image only, currently in this PR we have a `SolutionRestuls` here to contain all the info returned by each solution, easier for users to access some info like counting numbers in counting solution. Also be consistent with our ultralytics `Results`.
- Some docs improvements to be more straightforward to users.
- Also updated some docs macro to a programmable form: https://github.com/ultralytics/ultralytics/pull/18491/files#diff-c819e1cdf35fd10e19e6f9ca25de171f95cc59bd1a45e27c0d1345c0fa9840a1, in this way we could refer some rows of the table by the argument name, instead of having all the elements. The benefit of this is that we could just have one `solutions-args.md` as macro containing all the arguments for every solution and distribute specific arguments to its page by explicitly passing the argument names, for example here's a example to access the arguments for `object-counting` solution.
    ```md
    {% from "macros/solutions-args.md" import param_table %}
    {{ param_table(["model", "show_in", "show_out", "region"]) }}
    ```

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
Fix Solutions CLI and docs

RizwanMunawar committed:
Merge 674b5ed5795e350a2c427755f6f659ae624208e6 into 13c5bd5b92cc974aeed432a09b47f2fcb3411040

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
revert solutions short name from docs

Laughing-q committed:
remove MAPPINGS

glenn-jocher commented:
@RizwanMunawar new conflicts here I think due to the recent updates to solutions init. Can you resolve these please? Thank you!

glenn-jocher commented:
@RizwanMunawar also I think the ruff reformat is not fixing comments on commented lines. All comments should be start with `[space, space, #, space, comment]` but I see many comments here with arbitrary spaces. Can you please fix these? 

One way to easily fix is to uncomment lines, run Ruff, then recomment.

glenn-jocher commented on docs/en/guides/analytics.md:
```diff
@@ -62,22 +62,22 @@ This guide provides a comprehensive overview of three fundamental types of [data
-            model="yolo11n.pt",  # Path to the YOLO11 model file
-            # classes=[0, 2],  # If you want to count specific classes i.e person and car with COCO pretrained model.
+            show=True,  # display the output
+            analytics_type="line",  # pass the analytics type, could be "pie", "bar" or "area".
+            model="yolo11n.pt",  # path to the YOLO11 model file
+            # classes=[0, 2],           # display analytics for specific detection classes
```
Example comment here on L80 that should have two spaces (like the ones on L78-79)

RizwanMunawar replied:
@glenn-jocher, I have tried to resolve all solutions docs pages issues here: [eaa5abe](https://github.com/ultralytics/ultralytics/pull/18491/commits/eaa5abe2d9074d619f56750a2957d570b240a6b8)

RizwanMunawar commented:
@glenn-jocher I will do that early morning tomorrow, as currently I am away from the system. In the meantime, you can review it? In the morning I can work on your recommendations. What do you think?

RizwanMunawar committed:
fix __init_.py conflicts

RizwanMunawar committed:
fix __init_.py conflicts

RizwanMunawar committed:
fix __init_.py conflicts

RizwanMunawar committed:
fix __init_.py conflicts

RizwanMunawar committed:
fix __init_.py conflicts

RizwanMunawar committed:
fix __init_.py conflicts

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
update __init__.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update parking_management.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update parking_management.py

RizwanMunawar committed:
update parking-management.md

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update mkdocs.yml

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

RizwanMunawar committed:
fix indentation issues

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
Merge branch 'main' into annotator-overrides-for-solutions

glenn-jocher committed:
Merge branch 'main' into annotator-overrides-for-solutions

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher commented:
@RizwanMunawar see failing Doc tests here with Ruff errors:
https://github.com/ultralytics/ultralytics/actions/runs/13773210233/job/38516412390

```
tests/test_solutions.py:100:13: F821 Undefined name `YOLO`
    |
 98 |     from ultralytics.utils.plotting import Annotator, colors
 99 |
100 |     model = YOLO(WEIGHTS_DIR / "yolo11n-seg.pt")
    |             ^^^^ F821
101 |     names = model.names
102 |     cap = cv2.VideoCapture(TMP / DEMO_VIDEO)
    |

tests/test_solutions.py:100:[18](https://github.com/ultralytics/ultralytics/actions/runs/13773210233/job/38516412390?pr=18491#step:8:19): F821 Undefined name `WEIGHTS_DIR`
    |
 98 |     from ultralytics.utils.plotting import Annotator, colors
 99 |
100 |     model = YOLO(WEIGHTS_DIR / "yolo11n-seg.pt")
    |                  ^^^^^^^^^^^ F821
101 |     names = model.names
102 |     cap = cv2.VideoCapture(TMP / DEMO_VIDEO)
    |

ultralytics/cfg/__init__.py:718:52: F821 Undefined name `s_n`
    |
716 |         from ultralytics import solutions
717 |
718 |         solution = getattr(solutions, SOLUTION_MAP[s_n])(
    |                                                    ^^^ F821
7[19](https://github.com/ultralytics/ultralytics/actions/runs/13773210233/job/38516412390?pr=18491#step:8:20) |             IS_CLI=True, **overrides
720 |         )  # get solution class i.e ObjectCounter
    |

ultralytics/cfg/__init__.py:740:62: F821 Undefined name `s_n`
    |
738 |                 if not success:
739 |                     break
740 |                 results = solution(frame, f_n := f_n + 1) if s_n == "analytics" else solution(frame)
    |                                                              ^^^ F821
741 |                 LOGGER.info(f"🚀 Results: {results}")
742 |                 if s_n != "crop":
    |

ultralytics/cfg/__init__.py:742:[20](https://github.com/ultralytics/ultralytics/actions/runs/13773210233/job/38516412390?pr=18491#step:8:21): F821 Undefined name `s_n`
    |
740 |                 results = solution(frame, f_n := f_n + 1) if s_n == "analytics" else solution(frame)
741 |                 LOGGER.info(f"🚀 Results: {results}")
742 |                 if s_n != "crop":
    |                    ^^^ F8[21](https://github.com/ultralytics/ultralytics/actions/runs/13773210233/job/38516412390?pr=18491#step:8:22)
743 |                     vw.write(results.plot_im)
744 |                 if cv2.waitKey(1) & 0xFF == ord("q"):
    |

ultralytics/solutions/heatmap.py:115:33: F821 Undefined name `plot_im`
    |
114 |         if self.region is not None:
115 |             self.display_counts(plot_im)  # Display the counts on the frame
    |                                 ^^^^^^^ F821
116 |
117 |         plot_im = self.annotator.result()
    |

Found 6 errors.
```

UltralyticsAssistant committed:
Merge branch 'main' into annotator-overrides-for-solutions

RizwanMunawar committed:
update test_solutions.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update __init__.py

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

RizwanMunawar committed:
update heatmap.py

RizwanMunawar committed:
update test_solutions.py

RizwanMunawar committed:
update test_solutions.py

RizwanMunawar committed:
update test_solutions.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update test_solutions.py

RizwanMunawar committed:
update test_solutions.py

RizwanMunawar committed:
fix functions name and update solutions.py

RizwanMunawar committed:
Merge d2ce7806cabae198f940a9feb4e33cf70f2166e2 into 40a8f5d0190dfefb6a42417c354f78d095a8ba30

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update solutions.py

RizwanMunawar committed:
Merge branch 'annotator-overrides-for-solutions' of https://github.com/ultralytics/ultralytics into annotator-overrides-for-solutions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar commented:
@glenn-jocher all the above docs issues are fixed! I have performed testing with solutions tests locally on MacOS, Linux and Windows, and everything working. Thanks

@Laughing-q FYI, [here](https://github.com/ultralytics/ultralytics/pull/18491/commits/d2ce7806cabae198f940a9feb4e33cf70f2166e2), I have improved the segmentation mask plotting a bit for better visualization. (My first idea was to use the mask plotting function we already have for segmentation, but then I noticed, it's not easy to use as we are doing some post-processing on [masks](https://github.com/ultralytics/ultralytics/blob/40a8f5d0190dfefb6a42417c354f78d095a8ba30/ultralytics/models/yolo/segment/predict.py#L52), before passing it to the plotting function, which I believe is slow in terms of single mask plotting). Thanks

glenn-jocher committed:
Merge branch 'main' into annotator-overrides-for-solutions

glenn-jocher committed:
Cleanup

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher committed:
Cleanup

glenn-jocher committed:
Merge remote-tracking branch 'origin/annotator-overrides-for-solutions' into annotator-overrides-for-solutions

glenn-jocher committed:
Cleanup

glenn-jocher committed:
Cleanup

glenn-jocher committed:
Cleanup

glenn-jocher committed:
Merge branch 'main' into annotator-overrides-for-solutions

glenn-jocher committed:
Update __init__.py

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher committed:
Cleanup

glenn-jocher committed:
Merge remote-tracking branch 'origin/annotator-overrides-for-solutions' into annotator-overrides-for-solutions

RizwanMunawar committed:
Merge ff495a64cfebc55434715b0cff3d1e00ee8cd53a into 5a26107620dcb1f606e60a0b33c56d1311aa72d5

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions
